File: rdkit-sys/wrapper/src/periodic_table.cc
#include "rust/cxx.h"
#include <rdkit/GraphMol/PeriodicTable.h>

namespace RDKit {
std::unique_ptr<PeriodicTable> get_periodic_table() {
	// in rdkit we have the following:
	// PeriodicTable::getTable() return : const PeriodicTable*
	// so we need to take into account that it is a const pointer
	return std::make_unique<PeriodicTable>(*PeriodicTable::getTable());
}
rust::String getElementSymbol(unsigned int atomic_number) {
	PeriodicTable *pt = RDKit::PeriodicTable::getTable();
	return pt->getElementSymbol(atomic_number);
}

rust::String getElementName(unsigned int atomic_number) {
	PeriodicTable *pt = RDKit::PeriodicTable::getTable();
	return pt->getElementName(atomic_number);
}

const std::vector<int> &getValenceList(unsigned int atomic_number) {
	PeriodicTable *pt = RDKit::PeriodicTable::getTable();
	return pt->getValenceList(atomic_number);
}

} // namespace RDKit

File: rdkit-sys/src/bridge/periodic_table.rs
use cxx::{CxxVector, UniquePtr};
use ffi::PeriodicTable;

#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        type PeriodicTable;
        include!("wrapper/include/periodic_table.h");
        pub fn get_periodic_table() -> UniquePtr<PeriodicTable>;
        pub fn getAtomicWeight(self: &PeriodicTable, atomic_number: u32) -> f64;
        pub fn getAtomicNumber(self: &PeriodicTable, atom: &CxxString) -> i32;
        pub fn getElementSymbol(atomic_number: u32) -> String;
        pub fn getElementName(atomic_number: u32) -> String;
        pub fn getRvdw(self: &PeriodicTable, atomic_number: u32) -> f64;
        pub fn getRcovalent(self: &PeriodicTable, atomic_number: u32) -> f64;
        pub fn getRb0(self: &PeriodicTable, atomic_number: u32) -> f64;
        pub fn getDefaultValence(self: &PeriodicTable, atomic_number: u32) -> i32;
        fn getValenceList(atomic_number: u32) -> &'static CxxVector<i32>;
        pub fn getNouterElecs(self: &PeriodicTable, atomic_number: u32) -> i32;
        pub fn getMostCommonIsotope(self: &PeriodicTable, atomic_number: u32) -> i32;
        pub fn getMostCommonIsotopeMass(self: &PeriodicTable, atom: &CxxString) -> f64;
    }
}

pub trait PeriodicTableOps {
    fn getElementSymbol(self, atomic_number: u32) -> String;
    fn getElementName(self, atomic_number: u32) -> String;
    fn getValenceList(self, atomic_number: u32) -> &'static CxxVector<i32>;
}
impl<'a> PeriodicTableOps for UniquePtr<PeriodicTable> {
    fn getElementSymbol(self, atomic_number: u32) -> String {
        ffi::getElementSymbol(atomic_number)
    }

    fn getElementName(self, atomic_number: u32) -> String {
        ffi::getElementName(atomic_number)
    }

    fn getValenceList(self, atomic_number: u32) -> &'static CxxVector<i32> {
        ffi::getValenceList(atomic_number)
    }
}
